tracer/src/Datadog.Trace/Telemetry/Collectors/MetricsTelemetryCollector.cs

```diff
@@ -0,0 +1,478 @@
+                            type: TelemetryMetricType.Count) { Tags = helper.GetTags(), });
+                }
+            }
+        }
+
+        return data;
+    }
+
+    private static List<DistributionMetricData>? GetDistributionData(
+        BoundedConcurrentQueue<double>[] distributions,
+        ConcurrentDictionary<MetricKey, BoundedConcurrentQueue<double>> distributionsWithTags)
+    {
+        var distributionsLength = distributions.Count(x => x.Count > 0);
+        var distributionsWithTagsLength = distributionsWithTags.Count(x => x.Value.Count > 0);
+
+        var totalLength = distributionsLength + distributionsWithTagsLength;
+        if (totalLength == 0)
+        {
+            return null;
+        }
+
+        var data = new List<DistributionMetricData>(totalLength);
+
+        if (distributionsLength > 0)
+        {
+            for (var i = 0; i < distributions.Length; i++)
+            {
+                var queue = distributions[i];
+                var metric = (Distribution)i;
+                if (queue.Count > 0 && metric.GetName() is { } metricName)
+                {
+                    data.Add(
+                        new DistributionMetricData(
+                            metricName,
+                            points: queue.ToArray(),
+                            common: metric.IsCommon()));
+                }
+            }
+        }
+
+        foreach (var kvp in distributionsWithTags)
+        {
+            var helper = kvp.Key;
+            var metric = (Distribution)helper.Metric;
+            if (metric.GetName() is { } metricName)
+            {
+                data.Add(
+                    new DistributionMetricData(
+                        metricName,
+                        points: kvp.Value.ToArray(),
+                        common: metric.IsCommon()) { Tags = helper.GetTags(), });
+            }
+        }
+
+        return data;
+    }
+
+    private void RecordCount(in MetricKey key, int increment)
+    {
+#if NETCOREAPP
+        // we can avoid the closure in .NET Core
+        _buffer.CountsWithTags.AddOrUpdate(
+            key: key,
+            addValueFactory: static (_, arg) => arg,
+            updateValueFactory: static (_, old, arg) => (old + arg) < old ? int.MaxValue : (old + arg),
+            factoryArgument: increment);
+#else
+        _buffer.CountsWithTags.AddOrUpdate(
+            key,
+            increment,
+            (_, old) => (old + increment) < old ? int.MaxValue : (old + increment)); // look out for overflow
+#endif
+    }
+
+    private void RecordGauge(in MetricKey key, int value)
+    {
+#if NETCOREAPP
+        // we can avoid the closure in .NET Core
+        _buffer.GaugesWithTags.AddOrUpdate(
+            key: key,
+            addValueFactory: static (_, arg) => arg,
+            updateValueFactory: static (_, _, arg) => arg,
+            factoryArgument: value);
+#else
+        _buffer.GaugesWithTags.AddOrUpdate(key, value, (_, old) => value);
+#endif
```

Comment:
You're overthinking it.

```suggestion
        _buffer.GaugesWithTags[key] = value;
```

Reply:
Quite. In fact, it was more like copy pasta from the `RecordCount` implementations 😉